# __Zepto Inventory Analysis__

In [ ]:
#Importing libraries
import pandas as pd
from sqlalchemy import create_engine

#Load CSV
df = pd.read_csv("C:/Users/souna/Downloads/archive/zepto_v2.csv", encoding='latin1')

#Data Cleaning
#--Convert paise to rupees
df['mrp'] = df['mrp'] / 100
df['discountedSellingPrice'] = df['discountedSellingPrice'] / 100

#--Remove zero-value rows
df = df[(df['mrp'] > 0) & (df['discountedSellingPrice'] > 0)]


#Save to MySQL
engine = create_engine("mysql+mysqlconnector://root:iamsounak71@localhost/zepto_sql_project")
df.to_sql('zepto_clean', con=engine, if_exists='replace', index=False)


3731

In [11]:
df.to_csv("C:/Users/souna/Downloads/zepto_clean.csv", index=False)

In [ ]:
%load_ext sql
%sql mysql+mysqlconnector://root:root@localhost/zepto_sql_project


In [7]:
import prettytable
prettytable.__dict__['DEFAULT'] = prettytable.MARKDOWN

C:\Users\souna\AppData\Local\Temp\ipykernel_11208\3940415966.py:2: DeprecationWarning: the 'MARKDOWN' constant is deprecated, use the 'TableStyle' enum instead
  prettytable.__dict__['DEFAULT'] = prettytable.MARKDOWN


In [20]:
%%sql

SELECT * 
FROM zepto_clean
LIMIT 10;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
10 rows affected.


[(1, 'Fruits & Vegetables', 'Onion', Decimal('25.00'), Decimal('16.00'), 3, Decimal('21.00'), 1000, 0, 1),
 (2, 'Fruits & Vegetables', 'Pomegranate Peeled', Decimal('58.00'), Decimal('15.00'), 2, Decimal('49.00'), 100, 0, 100),
 (3, 'Cooking Essentials', '24 Mantra Organic Red Poha', Decimal('75.00'), Decimal('8.00'), 1, Decimal('69.00'), 500, 0, 500),
 (4, 'Cooking Essentials', '24 Mantra Organic Sattu Atta', Decimal('95.00'), Decimal('0.00'), 6, Decimal('95.00'), 500, 0, 500),
 (5, 'Fruits & Vegetables', 'Tomato Hybrid', Decimal('42.00'), Decimal('16.00'), 3, Decimal('35.00'), 1000, 0, 1),
 (6, 'Fruits & Vegetables', 'Tender Coconut', Decimal('51.00'), Decimal('15.00'), 3, Decimal('43.00'), 58, 0, 1),
 (7, 'Fruits & Vegetables', 'Coriander Leaves', Decimal('20.00'), Decimal('15.00'), 3, Decimal('17.00'), 100, 0, 100),
 (8, 'Fruits & Vegetables', 'Ladies Finger ', Decimal('14.00'), Decimal('14.00'), 3, Decimal('12.00'), 250, 0, 250),
 (9, 'Fruits & Vegetables', 'Potato', Decimal('35.00'), Decimal('17.00'), 3, Decimal('29.00'), 1000, 0, 1),
 (10, 'Fruits & Vegetables', 'Lemon', Decimal('75.00'), Decimal('16.00'), 3, Decimal('63.00'), 200, 0, 200)]

## Database Exploration

### Sample Data

In [21]:
%%sql

SELECT * FROM zepto_clean LIMIT 10;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
10 rows affected.


[(1, 'Fruits & Vegetables', 'Onion', Decimal('25.00'), Decimal('16.00'), 3, Decimal('21.00'), 1000, 0, 1),
 (2, 'Fruits & Vegetables', 'Pomegranate Peeled', Decimal('58.00'), Decimal('15.00'), 2, Decimal('49.00'), 100, 0, 100),
 (3, 'Cooking Essentials', '24 Mantra Organic Red Poha', Decimal('75.00'), Decimal('8.00'), 1, Decimal('69.00'), 500, 0, 500),
 (4, 'Cooking Essentials', '24 Mantra Organic Sattu Atta', Decimal('95.00'), Decimal('0.00'), 6, Decimal('95.00'), 500, 0, 500),
 (5, 'Fruits & Vegetables', 'Tomato Hybrid', Decimal('42.00'), Decimal('16.00'), 3, Decimal('35.00'), 1000, 0, 1),
 (6, 'Fruits & Vegetables', 'Tender Coconut', Decimal('51.00'), Decimal('15.00'), 3, Decimal('43.00'), 58, 0, 1),
 (7, 'Fruits & Vegetables', 'Coriander Leaves', Decimal('20.00'), Decimal('15.00'), 3, Decimal('17.00'), 100, 0, 100),
 (8, 'Fruits & Vegetables', 'Ladies Finger ', Decimal('14.00'), Decimal('14.00'), 3, Decimal('12.00'), 250, 0, 250),
 (9, 'Fruits & Vegetables', 'Potato', Decimal('35.00'), Decimal('17.00'), 3, Decimal('29.00'), 1000, 0, 1),
 (10, 'Fruits & Vegetables', 'Lemon', Decimal('75.00'), Decimal('16.00'), 3, Decimal('63.00'), 200, 0, 200)]

### Count Of Rows

In [22]:
%%sql

SELECT COUNT(*) FROM zepto_clean;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
1 rows affected.


[(3731,)]

### All Columns from Table zepto_clean

In [23]:
%%sql

SELECT 
    COLUMN_NAME, 
    DATA_TYPE, 
    IS_NULLABLE, 
    CHARACTER_MAXIMUM_LENGTH
FROM INFORMATION_SCHEMA.COLUMNS
WHERE TABLE_NAME = 'zepto_clean';

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
10 rows affected.


[('availableQuantity', 'int', 'YES', None),
 ('Category', 'varchar', 'YES', 120),
 ('discountedSellingPrice', 'decimal', 'YES', None),
 ('discountPercent', 'decimal', 'YES', None),
 ('mrp', 'decimal', 'YES', None),
 ('name', 'varchar', 'YES', 150),
 ('outOfStock', 'tinyint', 'YES', None),
 ('quantity', 'int', 'YES', None),
 ('sku_id', 'int', 'NO', None),
 ('weightInGms', 'int', 'YES', None)]

### Null Values

In [24]:
%%sql

SELECT * FROM zepto_clean
WHERE name IS NULL
	OR Category IS NULL
	OR mrp IS NULL
	OR discountPercent IS NULL
	OR discountedSellingPrice IS NULL
	OR weightInGms IS NULL
	OR availableQuantity IS NULL
	OR outOfStock IS NULL
	OR quantity IS NULL;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
0 rows affected.


[]

### Different Product Categories

In [25]:
%%sql 

SELECT DISTINCT Category 
FROM zepto_clean
ORDER BY Category;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
14 rows affected.


[('Beverages',),
 ('Biscuits',),
 ('Chocolates & Candies',),
 ('Cooking Essentials',),
 ('Dairy, Bread & Batter',),
 ('Fruits & Vegetables',),
 ('Health & Hygiene',),
 ('Home & Cleaning',),
 ('Ice Cream & Desserts',),
 ('Meats, Fish & Eggs',),
 ('Munchies',),
 ('Paan Corner',),
 ('Packaged Food',),
 ('Personal Care',)]

###  Products per category

In [30]:
%%sql

SELECT Category, COUNT(*) AS total_products
FROM zepto_clean
GROUP BY Category
ORDER BY total_products DESC;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
14 rows affected.


[('Cooking Essentials', 514),
 ('Munchies', 514),
 ('Packaged Food', 388),
 ('Ice Cream & Desserts', 388),
 ('Chocolates & Candies', 388),
 ('Personal Care', 344),
 ('Paan Corner', 344),
 ('Home & Cleaning', 193),
 ('Biscuits', 147),
 ('Dairy, Bread & Batter', 129),
 ('Beverages', 129),
 ('Health & Hygiene', 97),
 ('Fruits & Vegetables', 93),
 ('Meats, Fish & Eggs', 63)]

### Products in stock vs out of stock

In [26]:
%%sql 

SELECT outOfStock, COUNT(*)
FROM zepto_clean
GROUP BY outOfStock;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
2 rows affected.


[(0, 3278), (1, 453)]

### Product names present multiple times

In [28]:
%%sql

SELECT name, COUNT(*) AS "Number of times"
FROM zepto_clean
GROUP BY name
HAVING COUNT(*) > 1
ORDER BY COUNT(*) DESC
LIMIT 10;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
10 rows affected.


[('Saffola Veggie Twist Masala Oats', 10),
 ('Arden Eggs White', 10),
 ("Mother's Recipe Tamarind Paste", 10),
 ('Quaker Oats', 10),
 ('Sunfeast Yippee! Magic Masala Noodles', 10),
 ('Sunfeast Yippee! Pasta Treat - Sour Cream Onion', 10),
 ('Amul Delicious Fat Spread - Cholesterol Free', 10),
 ("Kellogg's Real Almond & Honey Corn Flakes", 9),
 ('Amul Fresh Cream', 8),
 ('iD Idli & Dosa Batter ', 7)]

### Summary statistics for numeric columns

In [29]:
%%sql

SELECT 
    MIN(mrp) AS min_mrp,
    MAX(mrp) AS max_mrp,
    AVG(mrp) AS avg_mrp,
    MIN(discountPercent) AS min_discount,
    MAX(discountPercent) AS max_discount,
    AVG(discountPercent) AS avg_discount,
    MIN(weightInGms) AS min_weight,
    MAX(weightInGms) AS max_weight,
    AVG(weightInGms) AS avg_weight
FROM zepto_clean;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
1 rows affected.


[(Decimal('10.00'), Decimal('2600.00'), Decimal('156.843206'), Decimal('0.00'), Decimal('51.00'), Decimal('7.619137'), 0, 10000, Decimal('387.9276'))]

### Discount distribution

In [33]:
%%sql

SELECT discountPercent, COUNT(*) AS count_products
FROM zepto_clean
GROUP BY discountPercent
ORDER BY discountPercent DESC
LIMIT 10;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
10 rows affected.


[(Decimal('51.00'), 3),
 (Decimal('50.00'), 55),
 (Decimal('49.00'), 3),
 (Decimal('46.00'), 2),
 (Decimal('45.00'), 2),
 (Decimal('43.00'), 4),
 (Decimal('40.00'), 10),
 (Decimal('35.00'), 2),
 (Decimal('33.00'), 2),
 (Decimal('32.00'), 2)]

### MRP vs discountedSellingPrice comparison

In [35]:
%%sql

SELECT name, mrp, discountedSellingPrice, (mrp - discountedSellingPrice) AS discount_amount
FROM zepto_clean
ORDER BY discount_amount DESC
LIMIT 15;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
15 rows affected.


[('Borges Extra Light Olive Oil Bottle ', Decimal('2600.00'), Decimal('1399.00'), Decimal('1201.00')),
 ('Borges Extra Light Olive Oil Bottle ', Decimal('2600.00'), Decimal('1399.00'), Decimal('1201.00')),
 ('Pampers Pants - Large', Decimal('1549.00'), Decimal('849.00'), Decimal('700.00')),
 ('Pampers Pants - Large', Decimal('1549.00'), Decimal('849.00'), Decimal('700.00')),
 ('Kelloggs Chocos - Moons & Stars', Decimal('660.00'), Decimal('396.00'), Decimal('264.00')),
 ('Kelloggs Chocos - Moons & Stars', Decimal('660.00'), Decimal('396.00'), Decimal('264.00')),
 ('Kelloggs Chocos - Moons & Stars', Decimal('660.00'), Decimal('396.00'), Decimal('264.00')),
 ('Pampers Active Baby Diapers New Born Extra Small ', Decimal('1099.00'), Decimal('849.00'), Decimal('250.00')),
 ('Pampers Active Baby Diapers New Born Extra Small ', Decimal('1099.00'), Decimal('849.00'), Decimal('250.00')),
 ('Popular Essentials Californian Almond', Decimal('695.00'), Decimal('470.00'), Decimal('225.00')),
 ('Popular Essentials Californian Almond', Decimal('695.00'), Decimal('470.00'), Decimal('225.00')),
 ('Dhara Health Refined Sun Flower Oil Jar', Decimal('1200.00'), Decimal('984.00'), Decimal('216.00')),
 ('Dhara Health Refined Sun Flower Oil Jar', Decimal('1200.00'), Decimal('984.00'), Decimal('216.00')),
 ('Pampers Premium Care Pants - Large', Decimal('1099.00'), Decimal('899.00'), Decimal('200.00')),
 ('Pampers Premium Care Pants - Large', Decimal('1099.00'), Decimal('899.00'), Decimal('200.00'))]

### Weight Distribution

In [37]:
%%sql

SELECT weightInGms, COUNT(*) AS count_products
FROM zepto_clean
GROUP BY weightInGms
ORDER BY weightInGms DESC
LIMIT 10;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
10 rows affected.


[(10000, 4),
 (5000, 42),
 (4000, 3),
 (3000, 12),
 (2000, 12),
 (1900, 1),
 (1500, 8),
 (1200, 15),
 (1160, 26),
 (1100, 3)]

## Data Analysis

#### The Questions:
1. Find the top 10 best-value products based on the discount percentage.
2. What are the Products with High MRP but Out of Stock
3. Calculate Estimated Revenue for each category
4. Find the top 5 categories generating the most estimated revenue.
5. What is the Total Inventory Weight Per Category
6. Find the Top 3 Most Expensive Products per Category
7. Find the Average Discount per Category and Show Beside Each Product
8. Show each product's percentage contribution to category revenue
9. Find all products where MRP is greater than ₹500 and discount is less than 10%.
10. Identify the top 5 categories offering the highest average discount percentage.
11. Find the price per gram for products above 100g and sort by best value.
12. Group the products into categories like Low, Medium, Bulk.
13. Which categories have the highest number of unique products?
14. Find categories with the highest average MRP per product.
15. Identify products with the highest markup (difference between MRP and discountedSellingPrice).
16. Find the total number of out-of-stock products per category.
17. Identify products with high discount but low stock (potential sell-out risk).
18. Find products with zero discount but still out of stock (high demand items)
19. Find the correlation between discount percentage and stock availability.

### Q1. Find the top 10 best-value products based on the discount percentage.

In [39]:
%%sql

SELECT DISTINCT name, mrp, discountPercent
FROM zepto_clean
ORDER BY discountPercent DESC
LIMIT 10;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
10 rows affected.


[('Dukes Waffy Chocolate Wafers', Decimal('45.00'), Decimal('51.00')),
 ('Dukes Waffy Orange Wafers', Decimal('45.00'), Decimal('51.00')),
 ('Dukes Waffy Strawberry Wafers', Decimal('45.00'), Decimal('51.00')),
 ('Ceres Foods Fish Mustard Instant Liquid Masala', Decimal('220.00'), Decimal('50.00')),
 ('Ceres Foods Laal Maas Instant Liquid Masala', Decimal('220.00'), Decimal('50.00')),
 ('Ceres Foods Nalli Nihari Instant Liquid Masala', Decimal('220.00'), Decimal('50.00')),
 ("Chef's Basket Durum Wheat Elbow Pasta", Decimal('160.00'), Decimal('50.00')),
 ("Chef's Basket Durum Wheat Fusilli Pasta", Decimal('160.00'), Decimal('50.00')),
 ("Chef's Basket Durum Wheat Penne Pasta", Decimal('160.00'), Decimal('50.00')),
 ('Dukes Waffy Chocolate Wafer Rolls ', Decimal('150.00'), Decimal('50.00'))]

#### Insights
- The best-value products are primarily from Dukes Waffy and Ceres Foods, each offering 50–51% discounts, indicating aggressive promotional pricing in snacks and instant masala categories.
- These discounts highlight price-sensitive marketing tactics in low-to-mid-priced packaged food items.

### Q2. What are the Products with High MRP but Out of Stock

In [40]:
%%sql

SELECT DISTINCT name, mrp
FROM zepto_clean
WHERE outOfStock = 1 and mrp > 250
ORDER BY mrp DESC;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
10 rows affected.


[("Patanjali Cow's Ghee", Decimal('565.00')),
 ('MamyPoko Pants Standard Diapers, Extra Large (12 - 17 kg)', Decimal('399.00')),
 ('Aashirvaad Atta With Mutigrains', Decimal('315.00')),
 ('Everest Kashmiri Lal Chilli Powder', Decimal('310.00')),
 ("Hershey's Cocoa + Almond Spread", Decimal('295.00')),
 ('Madhur Pure And Hygienic Sugar', Decimal('295.00')),
 ('RRO Mozzarella Block Cheese', Decimal('295.00')),
 ('Godrej Real Good Chicken Boneless Cubes ', Decimal('275.00')),
 ('Zorabian Chicken Cubes', Decimal('270.00')),
 ('Kelloggs Corn Flakes With Real Strawberry Pure', Decimal('265.00'))]

#### Insights
- Premium household items such as Patanjali Ghee, Aashirvaad Atta, and MamyPoko Diapers show frequent stockouts despite their high MRP. 
- Maintaining better inventory forecasts for these fast-moving, high-value products can help prevent missed sales opportunities.

### Q3. Calculate Estimated Revenue for each category

In [41]:
%%sql
SELECT Category, SUM(discountedSellingPrice * availableQuantity) AS total_revenue
FROM zepto_clean
GROUP BY Category
ORDER BY total_revenue DESC;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
14 rows affected.


[('Cooking Essentials', Decimal('337369.00')),
 ('Munchies', Decimal('337369.00')),
 ('Personal Care', Decimal('270849.00')),
 ('Paan Corner', Decimal('270849.00')),
 ('Packaged Food', Decimal('224385.00')),
 ('Ice Cream & Desserts', Decimal('224385.00')),
 ('Chocolates & Candies', Decimal('224385.00')),
 ('Home & Cleaning', Decimal('122661.00')),
 ('Health & Hygiene', Decimal('64180.00')),
 ('Dairy, Bread & Batter', Decimal('55051.00')),
 ('Beverages', Decimal('55051.00')),
 ('Biscuits', Decimal('25007.60')),
 ('Meats, Fish & Eggs', Decimal('20693.00')),
 ('Fruits & Vegetables', Decimal('10846.00'))]

#### Insights
- Cooking Essentials and Munchies contribute the highest estimated revenue, indicating they are the strongest-performing categories in terms of sales volume and pricing impact.
- Personal Care and Paan Corner also show strong performance, reflecting diversified demand. 
- Categories like Fruits & Vegetables and Meats, Fish & Eggs generate the least revenue, possibly due to perishable nature or lower margins. 
This insight helps prioritize restocking and marketing budgets toward high-revenue categories.

### Q4. Find the top 5 categories generating the most estimated revenue.

In [42]:
%%sql

SELECT Category, SUM(discountedSellingPrice * availableQuantity) AS total_revenue
FROM zepto_clean
GROUP BY Category
ORDER BY total_revenue DESC
LIMIT 5;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
5 rows affected.


[('Cooking Essentials', Decimal('337369.00')),
 ('Munchies', Decimal('337369.00')),
 ('Personal Care', Decimal('270849.00')),
 ('Paan Corner', Decimal('270849.00')),
 ('Packaged Food', Decimal('224385.00'))]

### Q5. What is the Total Inventory Weight Per Category

In [43]:
%%sql

SELECT Category, SUM(weightInGms * availableQuantity) AS total_weight
FROM zepto_clean
GROUP BY Category
ORDER BY total_weight DESC;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
14 rows affected.


[('Cooking Essentials', Decimal('1404654')),
 ('Munchies', Decimal('1404654')),
 ('Packaged Food', Decimal('490797')),
 ('Ice Cream & Desserts', Decimal('490797')),
 ('Chocolates & Candies', Decimal('490797')),
 ('Home & Cleaning', Decimal('373161')),
 ('Personal Care', Decimal('348187')),
 ('Paan Corner', Decimal('348187')),
 ('Dairy, Bread & Batter', Decimal('143735')),
 ('Beverages', Decimal('143735')),
 ('Health & Hygiene', Decimal('142904')),
 ('Fruits & Vegetables', Decimal('91794')),
 ('Biscuits', Decimal('84431')),
 ('Meats, Fish & Eggs', Decimal('48016'))]

#### Insights
Cooking Essentials and Munchies have the highest total inventory weight, reflecting large stock volumes required for high-demand categories.

### Q6. Find the Top 3 Most Expensive Products per Category

In [10]:
%%sql

SELECT *
FROM (
    SELECT 
        Category, name, mrp, ROW_NUMBER() OVER (PARTITION BY Category ORDER BY mrp DESC) AS rank_in_category
    FROM zepto_clean
) ranked
WHERE rank_in_category <= 3
ORDER BY Category, mrp DESC, name;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
42 rows affected.


[('Beverages', 'Amulya Dairy Whitener ', Decimal('380.00'), 1),
 ('Beverages', 'Amul Cheese Slices', Decimal('360.00'), 2),
 ('Beverages', 'RRO Mascarpone Cheese', Decimal('355.00'), 3),
 ('Biscuits', 'Danish Butter Cookies', Decimal('350.00'), 1),
 ('Biscuits', 'Britannia Good Day Cashew Cookies', Decimal('230.00'), 2),
 ('Biscuits', 'Karachi Bakery Cashew Biscuits ', Decimal('210.00'), 3),
 ('Chocolates & Candies', 'Pedigree Puppy Dry Dog Food Food Chicken & Milk', Decimal('690.00'), 1),
 ('Chocolates & Candies', 'Pedigree Dog Food Adult Chicken & Veg', Decimal('660.00'), 3),
 ('Chocolates & Candies', 'Pedigree Dog Food Adult Meat & Rice', Decimal('660.00'), 2),
 ('Cooking Essentials', 'Borges Extra Light Olive Oil Bottle ', Decimal('2600.00'), 1),
 ('Cooking Essentials', 'Praakritik Natural Desi Gir Cow A2 Ghee', Decimal('1450.00'), 2),
 ('Cooking Essentials', 'Dhara Kachi Ghani Mustard Oil Jar', Decimal('1250.00'), 3),
 ('Dairy, Bread & Batter', 'Amulya Dairy Whitener ', Decimal('380.00'), 1),
 ('Dairy, Bread & Batter', 'Amul Cheese Slices', Decimal('360.00'), 2),
 ('Dairy, Bread & Batter', 'RRO Mascarpone Cheese', Decimal('355.00'), 3),
 ('Fruits & Vegetables', 'Apple Royal Gala', Decimal('189.00'), 1),
 ('Fruits & Vegetables', 'Apple Washington', Decimal('177.00'), 2),
 ('Fruits & Vegetables', 'Avocado Indian Premium', Decimal('177.00'), 3),
 ('Health & Hygiene', 'Whisper Ultra Soft XL+ Sanitary Pads', Decimal('555.00'), 1),
 ('Health & Hygiene', 'Organic India Organic Chyawanprash', Decimal('495.00'), 2),
 ('Health & Hygiene', 'Whisper Ultra Clean Wings XL+  Pouch', Decimal('449.00'), 3),
 ('Home & Cleaning', 'Surf Excel Matic Powder Front Load ', Decimal('810.00'), 1),
 ('Home & Cleaning', 'Ariel Matic Front Load Detergent Washing Powder', Decimal('800.00'), 2),
 ('Home & Cleaning', 'Surf Excel Matic Top Load', Decimal('720.00'), 3),
 ('Ice Cream & Desserts', 'Pedigree Puppy Dry Dog Food Food Chicken & Milk', Decimal('690.00'), 1),
 ('Ice Cream & Desserts', 'Kelloggs Chocos - Moons & Stars', Decimal('660.00'), 3),
 ('Ice Cream & Desserts', 'Pedigree Dog Food Adult Chicken & Veg', Decimal('660.00'), 2),
 ('Meats, Fish & Eggs', 'Fresh Mutton Mince - Kheema (Halal)', Decimal('515.00'), 1),
 ('Meats, Fish & Eggs', 'Fresh Mutton Curry Cut Cubes (Halal)', Decimal('465.00'), 2),
 ('Meats, Fish & Eggs', 'TGFC Basa Fillets', Decimal('329.00'), 3),
 ('Munchies', 'Borges Extra Light Olive Oil Bottle ', Decimal('2600.00'), 1),
 ('Munchies', 'Praakritik Natural Desi Gir Cow A2 Ghee', Decimal('1450.00'), 2),
 ('Munchies', 'Dhara Kachi Ghani Mustard Oil Jar', Decimal('1250.00'), 3),
 ('Paan Corner', 'Pampers Pants - Large', Decimal('1549.00'), 1),
 ('Paan Corner', 'Pampers Premium Care Pants - Large', Decimal('1099.00'), 3),
 ('Paan Corner', 'Pampers Premium Care Pants - Medium', Decimal('1099.00'), 2),
 ('Packaged Food', 'Pedigree Puppy Dry Dog Food Food Chicken & Milk', Decimal('690.00'), 1),
 ('Packaged Food', 'Pedigree Dog Food Adult Chicken & Veg', Decimal('660.00'), 3),
 ('Packaged Food', 'Pedigree Dog Food Adult Meat & Rice', Decimal('660.00'), 2),
 ('Personal Care', 'Pampers Pants - Large', Decimal('1549.00'), 1),
 ('Personal Care', 'Pampers Active Baby Diapers New Born Extra Small ', Decimal('1099.00'), 2),
 ('Personal Care', 'Pampers Premium Care Pants - Medium', Decimal('1099.00'), 3)]

#### Insights
- The top 3 most expensive products per category are dominated by premium household items, baby care products, and specialty foods, showing clear price stratification across categories.
- Categories like Cooking Essentials, Munchies, and Personal Care feature high-value products such as premium oils, ghee, and diapers, while Fruits & Vegetables remain relatively low-priced. 
- This highlights the diversity in pricing strategy, with certain categories relying on high-ticket items for revenue, whereas others focus on volume sales. 
- Identifying these top-priced products can help in targeted marketing, premium promotions, and inventory prioritization.

### Q7. Find the Average Discount per Category and Show Beside Each Product

In [13]:
%%sql

SELECT category, name, discountPercent, ROUND(AVG(discountPercent) OVER (PARTITION BY category),2) AS avg_discount_category
FROM zepto_clean
ORDER BY category, discountPercent DESC
LIMIT 20;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
20 rows affected.


[('Beverages', 'RRO Fresh Ricotta', Decimal('50.00'), Decimal('7.16')),
 ('Beverages', 'RRO Mozzarella Pizza Cheese', Decimal('50.00'), Decimal('7.16')),
 ('Beverages', 'RRO Burrata Cheese', Decimal('50.00'), Decimal('7.16')),
 ('Beverages', 'RRO Cheddar Block Cheese', Decimal('50.00'), Decimal('7.16')),
 ('Beverages', 'RRO Sour Cream', Decimal('50.00'), Decimal('7.16')),
 ('Beverages', 'RRO Mascarpone Cheese', Decimal('50.00'), Decimal('7.16')),
 ('Beverages', 'RRO Mozzarella Block Cheese', Decimal('50.00'), Decimal('7.16')),
 ('Beverages', 'Suguna Fit Egg', Decimal('43.00'), Decimal('7.16')),
 ('Beverages', 'Epigamia Almond Milk Unsweetened Dairy Free', Decimal('33.00'), Decimal('7.16')),
 ('Beverages', 'RRO Goat Milk Fetta Cheee', Decimal('31.00'), Decimal('7.16')),
 ('Beverages', 'Dlicia Whip Topping Bottle', Decimal('27.00'), Decimal('7.16')),
 ('Beverages', 'The Whole Truth Dark Chocolate Peanut Butter - Creamy', Decimal('25.00'), Decimal('7.16')),
 ('Beverages', 'The Whole Truth Peanut Butter - Unsweetened & Crunchy', Decimal('20.00'), Decimal('7.16')),
 ('Beverages', 'The Whole Truth Peanut Butter - Sweetened & Crunchy', Decimal('20.00'), Decimal('7.16')),
 ('Beverages', 'The Whole Truth Dark Chocolate Peanut Butter - Crunchy', Decimal('20.00'), Decimal('7.16')),
 ('Beverages', 'Suguna Vitamin D Egg', Decimal('17.00'), Decimal('7.16')),
 ('Beverages', 'Suguna Nourish Egg ', Decimal('16.00'), Decimal('7.16')),
 ('Beverages', 'Arden Eggs White', Decimal('15.00'), Decimal('7.16')),
 ('Beverages', 'iD Idli & Dosa Batter ', Decimal('15.00'), Decimal('7.16')),
 ('Beverages', 'Suguna Active - DHA', Decimal('13.00'), Decimal('7.16'))]

### Q8. Show each product's percentage contribution to category revenue

In [15]:
%%sql

SELECT 
    Category, name, discountedSellingPrice * quantity AS Product_Revenue,
    SUM(discountedSellingPrice * quantity) OVER (PARTITION BY Category) AS Category_Total_Revenue,
    ROUND(
        (discountedSellingPrice * quantity) * 100.0 /
        SUM(discountedSellingPrice * quantity) OVER (PARTITION BY Category),
        2
    ) AS Percentage_Contribution
FROM zepto_clean
ORDER BY Category, Percentage_Contribution DESC
LIMIT 10;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
10 rows affected.


[('Beverages', 'Amul Cheese Slices', Decimal('269250.00'), Decimal('3330483.00'), Decimal('8.08')),
 ('Beverages', 'Amul Cheese - Block', Decimal('120000.00'), Decimal('3330483.00'), Decimal('3.60')),
 ('Beverages', 'Amul Butter - Pasteurised', Decimal('120000.00'), Decimal('3330483.00'), Decimal('3.60')),
 ('Beverages', 'Amul Unsalted Cooking Butter', Decimal('119000.00'), Decimal('3330483.00'), Decimal('3.57')),
 ('Beverages', 'Amul Safed Makkhan', Decimal('116000.00'), Decimal('3330483.00'), Decimal('3.48')),
 ('Beverages', 'Yoga Bar Peanut Butter Dark Chocolate Jar', Decimal('99600.00'), Decimal('3330483.00'), Decimal('2.99')),
 ('Beverages', 'Go Cheese Processed Block', Decimal('94800.00'), Decimal('3330483.00'), Decimal('2.85')),
 ('Beverages', 'Nestle Everyday Dairy Whitener', Decimal('84000.00'), Decimal('3330483.00'), Decimal('2.52')),
 ('Beverages', 'Yogabar 100% Pure Peanut Butter Creamy & Yummy Unsweetened', Decimal('82400.00'), Decimal('3330483.00'), Decimal('2.47')),
 ('Beverages', 'Epigamia Greek Yogurt - Natural', Decimal('80000.00'), Decimal('3330483.00'), Decimal('2.40'))]

#### Insights
- Most categories are revenue-skewed, with a few top products contributing a large share while the majority have smaller impacts. 
- This indicates dependency on key high-performing SKUs, highlighting opportunities to diversify sales across the product range.

### Q9. Find all products where MRP is greater than ₹500 and discount is less than 10%

In [16]:
%%sql

SELECT DISTINCT name, mrp, discountPercent
FROM zepto_clean
WHERE mrp > 500 AND discountPercent < 10
ORDER BY mrp DESC, discountPercent DESC;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
39 rows affected.


[('Dhara Kachi Ghani Mustard Oil Jar', Decimal('1250.00'), Decimal('8.00')),
 ('Saffola Gold (Jar)', Decimal('1240.00'), Decimal('0.00')),
 ('Dhara Filtered Groundnut Oil (Jar)', Decimal('1050.00'), Decimal('1.00')),
 ('Fortune Rice Bran Health Oil (Jar)', Decimal('1050.00'), Decimal('1.00')),
 ('Dhara Filtered Groundnut Oil (Jar)', Decimal('1050.00'), Decimal('0.00')),
 ('Fortune Soyabean Oil', Decimal('1005.00'), Decimal('0.00')),
 ('Fortune Sunlite Refined Sunflower (Jar)', Decimal('925.00'), Decimal('0.00')),
 ('Surf Excel Matic Powder Front Load ', Decimal('810.00'), Decimal('7.00')),
 ('Surf Excel Matic Top Load', Decimal('720.00'), Decimal('9.00')),
 ('Pedigree Puppy Dry Dog Food Food Chicken & Milk', Decimal('690.00'), Decimal('6.00')),
 ('Pedigree Dog Food Adult Meat & Rice', Decimal('660.00'), Decimal('7.00')),
 ('Lizol Double Concentrate Disinfectant Floor Cleaner Citrus', Decimal('650.00'), Decimal('8.00')),
 ('Nestle Nestle Nan Pro Follow-Up Formula', Decimal('650.00'), Decimal('0.00')),
 ('Dove Daily Shine Shampoo', Decimal('645.00'), Decimal('5.00')),
 ("L'Oreal Paris Excellence Creme Hair Color, 1 Black", Decimal('620.00'), Decimal('0.00')),
 ("L'Oreal Paris Excellence Creme Hair Color, 3 Dark Brown/Natural Darkest Brown", Decimal('620.00'), Decimal('0.00')),
 ("L'Oreal Paris Excellence Creme Hair Color, 4 Natural Brown/Natural Dark Brown", Decimal('620.00'), Decimal('0.00')),
 ("L'Oreal Paris Excellence Creme Hair Color, 4.25 Aishwarya's Brown", Decimal('620.00'), Decimal('0.00')),
 ('Nestle Nan Pro 4 Follow Up Formula Powder For Older Infants ', Decimal('620.00'), Decimal('0.00')),
 ('Pediasure Premium Chocolate Powdered Health Drink', Decimal('610.00'), Decimal('0.00')),
 ('Surf Excel Quick Wash', Decimal('600.00'), Decimal('8.00')),
 ('Gowardhan Ghee Jar', Decimal('598.00'), Decimal('3.00')),
 ('Organic India Apple Cider Vinegar', Decimal('595.00'), Decimal('9.00')),
 ("Bagrry's Fruit Fibre Muesli Mixed Fruit ", Decimal('595.00'), Decimal('0.00')),
 ('Bagrrys Fruit N Fibre Muesli - Strawberry', Decimal('595.00'), Decimal('0.00')),
 ("Patanjali Cow's Ghee", Decimal('565.00'), Decimal('0.00')),
 ('Sunsilk Lusciously Thick & Long Shampoo', Decimal('555.00'), Decimal('6.00')),
 ('Sunsilk Stunning Black Shine Shampoo', Decimal('555.00'), Decimal('6.00')),
 ('Whisper Ultra Soft XL+ Sanitary Pads', Decimal('555.00'), Decimal('0.00')),
 ("Kellogg's Real Almond & Honey Corn Flakes", Decimal('550.00'), Decimal('5.00')),
 ("L'Oreal Casting Creme Ebony Black Hair Color", Decimal('550.00'), Decimal('0.00')),
 ("L'Oreal Paris Casting Creme Gloss Hair Color (Dark Brown 400)", Decimal('550.00'), Decimal('0.00')),
 ("L'Oreal Paris Casting Creme Gloss Hair Color, Medium Brown 500", Decimal('550.00'), Decimal('0.00')),
 ("L'Oreal Paris Excellence Creme Hair Color, 3.16 Burgundy", Decimal('550.00'), Decimal('0.00')),
 ("L'Oreal Paris Shine Masque Royal Jelly Hair Colour", Decimal('550.00'), Decimal('0.00')),
 ('Surf Excel Matic Front Load Detergent Powder', Decimal('530.00'), Decimal('4.00')),
 ('Amul Ghee (Tin)', Decimal('520.00'), Decimal('2.00')),
 ('Fresh Mutton Mince - Kheema (Halal)', Decimal('515.00'), Decimal('3.00')),
 ('Amul Pure Ghee (Pouch)', Decimal('505.00'), Decimal('1.00'))]

#### Insights
High-value essentials like cooking oils, detergents, pet food, and personal care items show minimal discounting, suggesting stable demand and premium positioning.

### Q10. Identify the top 5 categories offering the highest average discount percentage

In [17]:
%%sql

SELECT Category, ROUND(AVG(discountPercent),2) AS avg_discount_percent
FROM zepto_clean
GROUP BY Category
ORDER BY avg_discount_percent DESC
LIMIT 5;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
5 rows affected.


[('Fruits & Vegetables', Decimal('15.46')),
 ('Meats, Fish & Eggs', Decimal('11.03')),
 ('Chocolates & Candies', Decimal('8.32')),
 ('Ice Cream & Desserts', Decimal('8.32')),
 ('Packaged Food', Decimal('8.32'))]

#### Insights
- Fruits & Vegetables and Meats, Fish & Eggs receive the highest average discounts, showing a push to clear perishable inventory quickly.
- These categories depend heavily on short-term promotions to maintain freshness and stimulate quick purchases, while stable packaged categories rely less on discounts for sales.

### Q11. Find the price per gram for products above 100g and sort by best value

In [27]:
%%sql

SELECT DISTINCT name, weightInGms, discountedSellingPrice, ROUND(discountedSellingPrice/weightInGms, 2) AS price_per_gm
FROM zepto_clean
WHERE weightInGms >= 100
ORDER BY price_per_gm
LIMIT 20;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
20 rows affected.


[('Aashirvaad Iodised Salt ', 1000, Decimal('19.00'), Decimal('0.02')),
 ('Onion', 1000, Decimal('21.00'), Decimal('0.02')),
 ('Onion', 3000, Decimal('57.00'), Decimal('0.02')),
 ('Shubh kart - Nirmal sugandhi mogra wet dhoop zipper 20 sticks', 1160, Decimal('28.00'), Decimal('0.02')),
 ('Tata Salt', 1000, Decimal('24.00'), Decimal('0.02')),
 ('Vicks Cough Drops Menthol', 1160, Decimal('20.00'), Decimal('0.02')),
 ('Baby Potato', 500, Decimal('16.00'), Decimal('0.03')),
 ('Beetroot', 500, Decimal('13.00'), Decimal('0.03')),
 ('Carrot', 500, Decimal('15.00'), Decimal('0.03')),
 ('Potato', 1000, Decimal('29.00'), Decimal('0.03')),
 ('Potato', 3000, Decimal('84.00'), Decimal('0.03')),
 ('Raw Banana', 500, Decimal('17.00'), Decimal('0.03')),
 ('Shubh kart - Tejas Twisted Cotton Wicks 1000n', 1000, Decimal('28.00'), Decimal('0.03')),
 ('Aashirvaad Atta', 10000, Decimal('399.00'), Decimal('0.04')),
 ('Beetroot', 250, Decimal('9.00'), Decimal('0.04')),
 ('Capsicum', 500, Decimal('21.00'), Decimal('0.04')),
 ('Carrot', 250, Decimal('11.00'), Decimal('0.04')),
 ('Pillsbury Chakki Fresh Atta', 10000, Decimal('375.00'), Decimal('0.04')),
 ('Radish White', 500, Decimal('19.00'), Decimal('0.04')),
 ('Tata Salt Lite', 1000, Decimal('36.00'), Decimal('0.04'))]

### Q12. Group the products into categories like Low, Medium, Bulk

In [57]:
%%sql

SELECT DISTINCT name, weightInGms,
	CASE 
		WHEN weightInGms <= 1000 THEN 'Low'
        WHEN weightInGms <= 5000 THEN 'Medium'
        ELSE 'Bulk'
        END AS weight_category
FROM zepto_clean
LIMIT 27 OFFSET 197;


 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
27 rows affected.


[('Dabur 100% Pure Honey - Worlds No.1', 1000, 'Low'),
 ('Aashirvaad Atta', 10000, 'Bulk'),
 ('Popular Essentials Star Anise', 25, 'Low'),
 ('Keya Wonder Red Hot Chilli Flakes', 40, 'Low'),
 ('Gits Medu Vada Instant Snack Mix', 200, 'Low'),
 ('Popular Essentials Toor Dal (Un-Polished)', 1000, 'Low'),
 ("Ching's Secret Chicken 65 Masala - Chicken", 20, 'Low'),
 ('Amul Ghee (Tin)', 1000, 'Low'),
 ('Weikfield Drinking Chocolate Powder', 100, 'Low'),
 ('Popular Essentials White Peas (Un-Polished)', 500, 'Low'),
 ('Popular Essentials Regular White Lobia', 500, 'Low'),
 ('Cadbury Cocoa Powder', 150, 'Low'),
 ('Puro Healthy Salt Crystal', 1000, 'Low'),
 ('Praakritik Natural Desi Gir Cow A2 Ghee', 540, 'Low'),
 ('Popular Essentials Premium Jeera Rice', 1000, 'Low'),
 ('Praakritik Organic Poha', 500, 'Low'),
 ('Amul Fresh Cream', 1000, 'Low'),
 ('Smith & Jones Shahi Paneer Masala Mix', 60, 'Low'),
 ('Praakritik Organic Natural Wild Forest Honey', 200, 'Low'),
 ('Popular Essentials Pure Jaggery', 450, 'Low'),
 ('Ching\x92s Secret Chowmein Hakka Noodles Masala ', 60, 'Low'),
 ('Popular Essentials Dry Dates (Kharak)', 250, 'Low'),
 ('Betty Crocker Mug Treats - Chocolate Mix - Pack of 4', 264, 'Low'),
 ('Aashirvaad Multigrain Atta', 1000, 'Low'),
 ('Popular Essentials Kashmiri Red Rajma', 500, 'Low'),
 ('Smith & Jones Pasta Masala', 108, 'Low'),
 ('Dhara Filtered Groundnut Oil (Jar)', 5000, 'Medium')]

### Q13. Which categories have the highest number of unique products?

In [20]:
%%sql

SELECT Category, COUNT(DISTINCT name) AS unique_products
FROM zepto_clean
GROUP BY Category
ORDER BY unique_products DESC;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
14 rows affected.


[('Cooking Essentials', 477),
 ('Munchies', 477),
 ('Chocolates & Candies', 365),
 ('Ice Cream & Desserts', 365),
 ('Packaged Food', 365),
 ('Paan Corner', 309),
 ('Personal Care', 309),
 ('Home & Cleaning', 183),
 ('Biscuits', 133),
 ('Beverages', 122),
 ('Dairy, Bread & Batter', 122),
 ('Fruits & Vegetables', 89),
 ('Health & Hygiene', 89),
 ('Meats, Fish & Eggs', 61)]

#### Insights
- Cooking Essentials and Munchies dominate in product variety, each offering 477 unique items, indicating these categories have the broadest selection for customers.
- Conversely, categories like Meats, Fish & Eggs have fewer unique products, reflecting limited SKUs for perishable or niche items.

### Q14. Find categories with the highest average MRP per product

In [21]:
%%sql

SELECT Category, ROUND(AVG(mrp), 2) AS avg_mrp
FROM zepto_clean
GROUP BY Category
ORDER BY avg_mrp DESC;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
14 rows affected.


[('Personal Care', Decimal('207.12')),
 ('Paan Corner', Decimal('207.12')),
 ('Meats, Fish & Eggs', Decimal('187.29')),
 ('Health & Hygiene', Decimal('159.18')),
 ('Cooking Essentials', Decimal('156.54')),
 ('Munchies', Decimal('156.54')),
 ('Packaged Food', Decimal('155.73')),
 ('Ice Cream & Desserts', Decimal('155.73')),
 ('Chocolates & Candies', Decimal('155.73')),
 ('Home & Cleaning', Decimal('152.94')),
 ('Dairy, Bread & Batter', Decimal('118.71')),
 ('Beverages', Decimal('118.71')),
 ('Biscuits', Decimal('59.49')),
 ('Fruits & Vegetables', Decimal('47.01'))]

#### Insights
- Higher average MRP in Personal Care and Paan Corner suggests these categories focus on premium products or items with higher perceived value.
- Lower-priced categories like Fruits & Vegetables and Biscuits may rely more on volume sales and discounts.

### Q15. Identify products with the highest markup (difference between MRP and discountedSellingPrice)

In [22]:
%%sql

SELECT name, Category, mrp, discountedSellingPrice, (mrp - discountedSellingPrice) AS markup
FROM zepto_clean
ORDER BY markup DESC
LIMIT 10;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
10 rows affected.


[('Borges Extra Light Olive Oil Bottle ', 'Cooking Essentials', Decimal('2600.00'), Decimal('1399.00'), Decimal('1201.00')),
 ('Borges Extra Light Olive Oil Bottle ', 'Munchies', Decimal('2600.00'), Decimal('1399.00'), Decimal('1201.00')),
 ('Pampers Pants - Large', 'Personal Care', Decimal('1549.00'), Decimal('849.00'), Decimal('700.00')),
 ('Pampers Pants - Large', 'Paan Corner', Decimal('1549.00'), Decimal('849.00'), Decimal('700.00')),
 ('Kelloggs Chocos - Moons & Stars', 'Packaged Food', Decimal('660.00'), Decimal('396.00'), Decimal('264.00')),
 ('Kelloggs Chocos - Moons & Stars', 'Chocolates & Candies', Decimal('660.00'), Decimal('396.00'), Decimal('264.00')),
 ('Kelloggs Chocos - Moons & Stars', 'Ice Cream & Desserts', Decimal('660.00'), Decimal('396.00'), Decimal('264.00')),
 ('Pampers Active Baby Diapers New Born Extra Small ', 'Personal Care', Decimal('1099.00'), Decimal('849.00'), Decimal('250.00')),
 ('Pampers Active Baby Diapers New Born Extra Small ', 'Paan Corner', Decimal('1099.00'), Decimal('849.00'), Decimal('250.00')),
 ('Popular Essentials Californian Almond', 'Cooking Essentials', Decimal('695.00'), Decimal('470.00'), Decimal('225.00'))]

#### Insights
Useful for analyzing how much price cushion or margin exists for products.

### Q16. Find the total number of out-of-stock products per category

In [23]:
%%sql

SELECT Category, COUNT(*) AS out_of_stock_count
FROM zepto_clean
WHERE outOfStock = '1'
GROUP BY Category
ORDER BY out_of_stock_count DESC;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
14 rows affected.


[('Cooking Essentials', 64),
 ('Munchies', 64),
 ('Packaged Food', 45),
 ('Ice Cream & Desserts', 45),
 ('Chocolates & Candies', 45),
 ('Biscuits', 42),
 ('Dairy, Bread & Batter', 28),
 ('Beverages', 28),
 ('Personal Care', 21),
 ('Paan Corner', 21),
 ('Home & Cleaning', 19),
 ('Health & Hygiene', 13),
 ('Meats, Fish & Eggs', 12),
 ('Fruits & Vegetables', 6)]

#### Insights
- The higher out-of-stock count in Cooking Essentials and Munchies could point to high demand, supply chain challenges, or inventory management issues. Businesses might need to optimize replenishment strategies for these categories to avoid lost sales. 
- Conversely, categories like Fruits & Vegetables and Meats, Fish & Eggs have fewer out-of-stock items, possibly due to smaller inventories or more controlled supply, suggesting steadier availability for customers.

### Q17. Identify products with high discount but low stock (potential sell-out risk)

In [24]:
%%sql

SELECT name, Category, discountPercent, availableQuantity
FROM zepto_clean
WHERE discountPercent > 40 AND availableQuantity < 7
ORDER BY discountPercent DESC;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
69 rows affected.


[('Dukes Waffy Chocolate Wafers', 'Biscuits', Decimal('51.00'), 4),
 ('Dukes Waffy Strawberry Wafers', 'Biscuits', Decimal('51.00'), 2),
 ('Dukes Waffy Orange Wafers', 'Biscuits', Decimal('51.00'), 5),
 ("Chef's Basket Durum Wheat Elbow Pasta", 'Cooking Essentials', Decimal('50.00'), 6),
 ("Chef's Basket Durum Wheat Fusilli Pasta", 'Cooking Essentials', Decimal('50.00'), 6),
 ('Ceres Foods Laal Maas Instant Liquid Masala', 'Cooking Essentials', Decimal('50.00'), 5),
 ('Ceres Foods Nalli Nihari Instant Liquid Masala', 'Cooking Essentials', Decimal('50.00'), 6),
 ("Chef's Basket Durum Wheat Penne Pasta", 'Cooking Essentials', Decimal('50.00'), 0),
 ('Ceres Foods Fish Mustard Instant Liquid Masala', 'Cooking Essentials', Decimal('50.00'), 6),
 ('Epigamia Fruit Yogurt Strawberry', 'Health & Hygiene', Decimal('50.00'), 0),
 ("Chef's Basket Durum Wheat Elbow Pasta", 'Munchies', Decimal('50.00'), 6),
 ("Chef's Basket Durum Wheat Fusilli Pasta", 'Munchies', Decimal('50.00'), 6),
 ('Ceres Foods Fish Mustard Instant Liquid Masala', 'Munchies', Decimal('50.00'), 6),
 ("Chef's Basket Durum Wheat Penne Pasta", 'Munchies', Decimal('50.00'), 0),
 ('RRO Cheddar Block Cheese', 'Dairy, Bread & Batter', Decimal('50.00'), 6),
 ('Epigamia Fruit Yogurt Alphonso Mango', 'Health & Hygiene', Decimal('50.00'), 6),
 ('RRO Fresh Ricotta', 'Dairy, Bread & Batter', Decimal('50.00'), 6),
 ('RRO Mozzarella Pizza Cheese', 'Dairy, Bread & Batter', Decimal('50.00'), 6),
 ('RRO Burrata Cheese', 'Dairy, Bread & Batter', Decimal('50.00'), 4),
 ('RRO Mascarpone Cheese', 'Dairy, Bread & Batter', Decimal('50.00'), 6),
 ('RRO Mozzarella Block Cheese', 'Dairy, Bread & Batter', Decimal('50.00'), 0),
 ('Ceres Foods Nalli Nihari Instant Liquid Masala', 'Munchies', Decimal('50.00'), 6),
 ('RRO Sour Cream', 'Dairy, Bread & Batter', Decimal('50.00'), 4),
 ('Ceres Foods Laal Maas Instant Liquid Masala', 'Munchies', Decimal('50.00'), 5),
 ('RRO Cheddar Block Cheese', 'Beverages', Decimal('50.00'), 6),
 ('RRO Fresh Ricotta', 'Beverages', Decimal('50.00'), 6),
 ('RRO Mozzarella Pizza Cheese', 'Beverages', Decimal('50.00'), 6),
 ('RRO Burrata Cheese', 'Beverages', Decimal('50.00'), 4),
 ('RRO Mascarpone Cheese', 'Beverages', Decimal('50.00'), 6),
 ('RRO Sour Cream', 'Beverages', Decimal('50.00'), 4),
 ("Chef's Basket Durum Wheat Fusilli Pasta", 'Packaged Food', Decimal('50.00'), 6),
 ('Moi Soi Sichuan Chilli Oil- For Stir Fry Marinade Spread & Dip', 'Packaged Food', Decimal('50.00'), 5),
 ('RRO Mozzarella Block Cheese', 'Beverages', Decimal('50.00'), 0),
 ('Moi Soi Manchurian Sauce - For Stir Fry Marinade Spread & Dip', 'Packaged Food', Decimal('50.00'), 6),
 ('Moi Soi Kung Pao Sauce - For Stir Fry Marinade Spread & Dip', 'Packaged Food', Decimal('50.00'), 5),
 ("Chef's Basket Durum Wheat Elbow Pasta", 'Packaged Food', Decimal('50.00'), 6),
 ('Moi Soi  Black Bean Sauce - Dip Spread Stir Fry & Marinade', 'Packaged Food', Decimal('50.00'), 5),
 ("Chef's Basket Durum Wheat Elbow Pasta", 'Ice Cream & Desserts', Decimal('50.00'), 6),
 ("Chef's Basket Durum Wheat Penne Pasta", 'Packaged Food', Decimal('50.00'), 0),
 ("Chef's Basket Durum Wheat Fusilli Pasta", 'Ice Cream & Desserts', Decimal('50.00'), 6),
 ('Moi Soi  Black Bean Sauce - Dip Spread Stir Fry & Marinade', 'Ice Cream & Desserts', Decimal('50.00'), 5),
 ('Moi Soi Sichuan Chilli Oil- For Stir Fry Marinade Spread & Dip', 'Ice Cream & Desserts', Decimal('50.00'), 5),
 ('Moi Soi Manchurian Sauce - For Stir Fry Marinade Spread & Dip', 'Ice Cream & Desserts', Decimal('50.00'), 6),
 ("Chef's Basket Durum Wheat Elbow Pasta", 'Chocolates & Candies', Decimal('50.00'), 6),
 ("Chef's Basket Durum Wheat Penne Pasta", 'Ice Cream & Desserts', Decimal('50.00'), 0),
 ('Moi Soi Kung Pao Sauce - For Stir Fry Marinade Spread & Dip', 'Ice Cream & Desserts', Decimal('50.00'), 5),
 ("Chef's Basket Durum Wheat Fusilli Pasta", 'Chocolates & Candies', Decimal('50.00'), 6),
 ('Moi Soi Kung Pao Sauce - For Stir Fry Marinade Spread & Dip', 'Chocolates & 

#### Insights
- Products like Dukes Waffy Wafers, Chef's Basket Pasta varieties, and Wingreens Cheesy Chipotle Sauce have both high discounts (≥50%) and very low stock (often ≤6 units), indicating a potential sell-out risk in the near term.
- High discounts are likely to drive demand, but low stock levels in these categories (Biscuits, Cooking Essentials, Packaged Food, Chocolates & Candies) mean these products could quickly go out of stock.
- These products may need restocking or price adjustments soon.

### Q18. Find products with zero discount but still out of stock (high demand items)

In [58]:
%%sql

SELECT name, Category, mrp, discountedSellingPrice
FROM zepto_clean
WHERE discountPercent = 0 AND outOfStock = 0
ORDER BY discountedSellingPrice DESC
LIMIT 20;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
20 rows affected.


[('Saffola Gold (Jar)', 'Munchies', Decimal('1240.00'), Decimal('1240.00')),
 ('Saffola Gold (Jar)', 'Cooking Essentials', Decimal('1240.00'), Decimal('1240.00')),
 ('Dhara Filtered Groundnut Oil (Jar)', 'Munchies', Decimal('1050.00'), Decimal('1050.00')),
 ('Dhara Filtered Groundnut Oil (Jar)', 'Cooking Essentials', Decimal('1050.00'), Decimal('1050.00')),
 ('Fortune Soyabean Oil', 'Cooking Essentials', Decimal('1005.00'), Decimal('999.00')),
 ('Fortune Soyabean Oil', 'Munchies', Decimal('1005.00'), Decimal('999.00')),
 ('Fortune Sunlite Refined Sunflower (Jar)', 'Cooking Essentials', Decimal('925.00'), Decimal('925.00')),
 ('Fortune Sunlite Refined Sunflower (Jar)', 'Munchies', Decimal('925.00'), Decimal('925.00')),
 ('Nestle Nestle Nan Pro Follow-Up Formula', 'Chocolates & Candies', Decimal('650.00'), Decimal('650.00')),
 ('Nestle Nestle Nan Pro Follow-Up Formula', 'Packaged Food', Decimal('650.00'), Decimal('650.00')),
 ('Nestle Nestle Nan Pro Follow-Up Formula', 'Ice Cream & Desserts', Decimal('650.00'), Decimal('650.00')),
 ("L'Oreal Paris Excellence Creme Hair Color, 3 Dark Brown/Natural Darkest Brown", 'Personal Care', Decimal('620.00'), Decimal('620.00')),
 ('Nestle Nan Pro 4 Follow Up Formula Powder For Older Infants ', 'Packaged Food', Decimal('620.00'), Decimal('620.00')),
 ('Nestle Nan Pro 4 Follow Up Formula Powder For Older Infants ', 'Chocolates & Candies', Decimal('620.00'), Decimal('620.00')),
 ("L'Oreal Paris Excellence Creme Hair Color, 1 Black", 'Paan Corner', Decimal('620.00'), Decimal('620.00')),
 ('Nestle Nan Pro 4 Follow Up Formula Powder For Older Infants ', 'Ice Cream & Desserts', Decimal('620.00'), Decimal('620.00')),
 ("L'Oreal Paris Excellence Creme Hair Color, 4.25 Aishwarya's Brown", 'Personal Care', Decimal('620.00'), Decimal('620.00')),
 ("L'Oreal Paris Excellence Creme Hair Color, 3 Dark Brown/Natural Darkest Brown", 'Paan Corner', Decimal('620.00'), Decimal('620.00')),
 ("L'Oreal Paris Excellence Creme Hair Color, 1 Black", 'Personal Care', Decimal('620.00'), Decimal('620.00')),
 ("L'Oreal Paris Excellence Creme Hair Color, 4 Natural Brown/Natural Dark Brown", 'Personal Care', Decimal('620.00'), Decimal('620.00'))]

#### Insights
These products are strong performers, people buy them even without discounts.

### Q19. Find the correlation between discount percentage and stock availability

In [26]:
%%sql

SELECT 
    CASE 
        WHEN discountPercent < 10 THEN 'Low Discount'
        WHEN discountPercent BETWEEN 10 AND 30 THEN 'Medium Discount'
        ELSE 'High Discount'
    END AS discount_range,
    AVG(CASE WHEN outOfStock = 0 THEN 1 ELSE 0 END) * 100 AS percent_out_of_stock
FROM zepto_clean
GROUP BY discount_range
ORDER BY percent_out_of_stock DESC;

 * mysql+mysqlconnector://root:***@localhost/zepto_sql_project
3 rows affected.


[('Medium Discount', Decimal('94.3839')),
 ('High Discount', Decimal('87.3563')),
 ('Low Discount', Decimal('84.7001'))]

#### Insights
- Checks if higher discounts correlate with higher stockouts.
- Products with Medium to High discounts have a higher percentage of being out of stock (Medium: 94.38%, High: 87.36%) compared to Low Discount products (84.70%). 
- This indicates that higher discount products are more likely to sell out or have lower stock on hand.